mc_rtc_rviz_panel/src/Schema.cpp

```diff
@@ -0,0 +1,319 @@
+#include "Schema.h"
+
+#include "SchemaArrayInput.h"
+
+#include <mc_rtc/logging.h>
+
+#include <boost/filesystem.hpp>
+namespace bfs = boost::filesystem;
+
+namespace mc_rtc_rviz
+{
+
+Schema::Schema(const std::string & file) : Schema(mc_rtc::Configuration{file}, file) {}
+
+Schema::Schema(const std::vector<Schema> & schemas)
+{
+  is_tuple_ = true;
+  create_form = [schemas](QWidget * parent, const mc_rtc::Configuration & data) {
+    std::vector<FormElement *> elems;
+    for(const auto & s : schemas)
+    {
+      for(auto & el : s.create_form(parent, data))
+      {
+        elems.push_back(el);
+      }
+    }
+    return elems;
+  };
+}
+
+Schema::Schema(const mc_rtc::Configuration & s, const std::string & source, const std::string & title, bool required_in)
+{
+  title_ = s("title", title.size() ? title : "No title property in " + source);
+  auto required = s("required", std::vector<std::string>{});
+  auto is_required = [&required](const std::string & label) {
+    return std::find(required.begin(), required.end(), label) != required.end();
+  };
+  /** Resolve a $ref entry into a Schema */
+  auto resolve_ref = [](const std::string & source, const std::string & ref_) -> const Schema & {
+    auto ref = ref_.substr(3); // remove leading /..
+    bfs::path ref_schema{source};
+    ref_schema = bfs::canonical(ref_schema.parent_path() / ref);
+    if(!store().count(ref_schema.string()))
+    {
+      store()[ref_schema.string()] = Schema{ref_schema.string()};
+    }
+    return store().at(ref_schema.string());
+  };
+  /** Handle enum entries */
+  auto handle_enum = [this](const std::string & k, bool required, const std::vector<std::string> & values) {
+    auto cf = create_form;
+    create_form = [cf, k, required, values](QWidget * parent, const mc_rtc::Configuration & data) {
+      auto v = cf(parent, data);
+      v.emplace_back(new form::ComboInput(parent, k, required, values, false));
+      if(values.size() == 1 && required)
+      {
+        v.back()->hidden(true);
+      }
+      return v;
+    };
+  };
+  /** Handle: boolean/integer/number/string */
+  auto handle_type = [this, &source](const std::string & k, bool required, const std::string & type,
+                                     const mc_rtc::Configuration & schema) {
+    auto cf = create_form;
+    if(type == "boolean")
+    {
+      create_form = [cf, k, required](QWidget * parent, const mc_rtc::Configuration & data) {
+        auto v = cf(parent, data);
+        v.emplace_back(new form::Checkbox(parent, k, required, true));
+        return v;
+      };
+    }
+    else if(type == "integer")
+    {
+      if(k == "robotIndex")
```

Comment:
There is a few places where we have `r1Index`/`r2Index` (`mc_rbdyn::Contact`, `CollisionConstraint`). Notably `AddRemoveContactTask` has an `mc_rbdyn::Contact`, and appears in the GUI. I haven't tried this PR yet, are those cases handled properly?